### PR TITLE
UI: add full graph toggle

### PR DESCRIPTION
**Pull request by jlukenoff**
### Problem

There are some reasonable scenarios where users would want to view only the direct successors and predecessors of the current selected node in the lineage graph. This PR enables users to toggle a "Full Graph" view, when deselected, the graph will render only the nodes that are along the selected paths. When deselected and changing selected nodes, we expect that the graph should re-render with any new nodes that weren't along the selected path previously.

Closes: #2558

### Solution

- Add a `showFullGraph` property to the lineage state and enables a toggle on the UI. 
- Add a `removeUnselectedNodes` method to `Lineage` that can remove any unselected nodes from the graph.

Screen Recording 2023-08-05 at 9 13 18 AM

One-line summary: add a toggle to the Lineage UI to let users switch between viewing the full graph and a view with only the selected paths.

### Checklist

- [x] You've [signed-off](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#sign-your-work) your work
- [ ] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've included a one-line summary of your change for the [`CHANGELOG.md`](https://github.com/MarquezProject/marquez/blob/main/CHANGELOG.md#unreleased) (_Depending on the change, this may not be necessary_).
- [ ] You've versioned your `.sql` database schema migration according to [Flyway's naming convention](https://flywaydb.org/documentation/concepts/migrations#naming) (_if relevant_)
- [x] You've included a [header](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#copyright--license) in any source code files (_if relevant_)

### CODE REVIEW

**Comment by davidsharp7** [comment]
Yeah looks good - only suggestion I would make is to change the background colour of the toggle to Marquez Green when its on so its a little easier to determine state.

**Comment by jlukenoff** [comment]
> Yeah looks good - only suggestion I would make is to change the background colour of the toggle to Marquez Green when its on so its a little easier to determine state.

Good idea, updated it to use the primary green color when enabled.

image

**Comment by davidsharp7** [comment]
> > Yeah looks good - only suggestion I would make is to change the background colour of the toggle to Marquez Green when its on so its a little easier to determine state.
> 
> Good idea, updated it to use the primary green color when enabled.
> 
> image

Yeah perfect!  Might need to get wider feedback but I wonder if its worth inverting the toggle so that by default its shows everything and activating the toggle shows only the critical path.

**Comment by jlukenoff** [comment]
That’s a good point, that may make more sense since the full graph has always been the default and this is something new users can enable. I’m having trouble thinking of a label for the switch that would make sense that isn’t too wordy though. Perhaps “Simple Graph” or something like that? 

Agree it would be great to get some wider feedback.

**Comment by davidsharp7** [comment]
> That’s a good point, that may make more sense since the full graph has always been the default and this is something new users can enable. I’m having trouble thinking of a label for the switch that would make sense that isn’t too wordy though. Perhaps “Simple Graph” or something like that?
> 
> Agree it would be great to get some wider feedback.

What about “Show Critical Path”?

**Comment by davidsharp7** [comment]
Yeah I think we can just reverse the logic and it should work fine

I've added the sections that can be reversed on your branch (could be wrong as this the first time I have ever looked at typescript/react/node)

File: web/src/store/reducers/lineage.ts
Comment (davidsharp7) [inline web/src/store/reducers/lineage.ts:30]:
```
showFullGraph: false
```

File: web/src/i18n/config.ts
Comment (davidsharp7) [inline web/src/i18n/config.ts:61]:
```
full_graph_label: 'Show Critical Path'
```

File: web/src/components/lineage/Lineage.tsx
Comment (davidsharp7) [inline web/src/components/lineage/Lineage.tsx:146]:
```
    if (this.props.selectedNode !== prevProps.selectedNode && this.props.showFullGraph) {
      this.buildGraphAll(this.props.lineage.graph)
    }
```

File: web/src/components/lineage/Lineage.tsx
Comment (davidsharp7) [inline web/src/components/lineage/Lineage.tsx:244]:
```
    if (this.props.showFullGraph) {
      this.removeUnselectedNodes()
    }
```

**Comment by davidsharp7** [comment]
Promise this is my last comment - there is a slight bug when you select "etl_delivery_7_days" - notice that there are 2 connected nodes even though they are not on the critical path.  I think its because technically you can't filter out those nodes because they are part of the critical path at some point.  I think a work around is to add something like this

```
    this.setState({
      graph: g,
      edges: this.props.showFullGraph ? this.getEdges().filter(edge => edge.isSelected === true) : this.getEdges(),
      nodes: g.nodes().map(v => g.node(v))
    })
```
Screenshot 2023-08-06 at 9 52 20 pm

**Comment by jlukenoff** [comment]
> Promise this is my last comment - there is a slight bug when you select "etl_delivery_7_days" - notice that there are 2 connected nodes even though they are not on the critical path. I think its because technically you can't filter out those nodes because they are part of the critical path at some point. I think a work around is to add something like this

I'm not sure I understand, doesn't the fact that they are connected by the selected path mean they are on the critical path?

File: web/src/i18n/config.ts
Comment (jlukenoff) [reply web/src/i18n/config.ts:61]: Eh, this just starts to feel like a lot of words to express something relatively simple. I tend to err on the side of "less is more" with these things. I really like the [ArgoCD](https://argo-cd.readthedocs.io/) version of this where the text labels show up as a tooltip when you hover over this options. I think that helps preserve the real-estate the configs take up. Maybe we want to think about something like that if we plan to add more configs to this part of the screen.

image

**Comment by jlukenoff** [comment]
Had to do a pretty big refactor to account for #2563 but the functionality it still the same. Plus it's worth it to inherit the react upgrade 😄

**Comment by davidsharp7** [comment]
> > Promise this is my last comment - there is a slight bug when you select "etl_delivery_7_days" - notice that there are 2 connected nodes even though they are not on the critical path. I think its because technically you can't filter out those nodes because they are part of the critical path at some point. I think a work around is to add something like this
> 
> I'm not sure I understand, doesn't the fact that they are connected by the selected path mean they are on the critical path? Could you point out which nodes in your screenshot you think shouldn't be displayed?

So if you select etl_delivery_7_days notice how there is a grey line between public.customers and email_discounts. Given the job selected that is not on the critical path and shouldn’t appear. Simplest solution is to also filter out anything where isSelected is False.

File: web/src/i18n/config.ts
Comment (davidsharp7) [reply web/src/i18n/config.ts:61]: Yeah I’m in two minds about that. Like the idea of having a pop up on hover to display basic ( maybe even facets like data quality?) info about the node. Maybe we need an additional menu on the header for global type variables like display and depth?

Comment (wslulciuc) [reply web/src/i18n/config.ts:61]: As a follow up, the lineage API should support the different graph options (`Full`, `Critical`) to avoid filtering on the UI. But, in the UI is a great first step!

**Comment by wslulciuc** [comment]
> So if you select etl_delivery_7_days notice how there is a grey line between public.customers and email_discounts. Given the job selected that is not on the critical path and shouldn’t appear. Simplest solution is to also filter out anything where isSelected is False.

@davidsharp7 I wanted to add to @jlukenoff response. In the `Full` graph mode, we'll want to show all dataset and job dependencies, but the connections between nodes in the graph don't need to be expanded by default. That is, they can be grouped and collapsed, allowing the user to discover and explore the graph as needed. For some insight into where I see the interaction of the graph going, see https://github.com/MarquezProject/marquez/issues/2014

File: web/src/__tests__/components/Lineage.test.tsx
Comment (phixMe) [inline web/src/__tests__/components/Lineage.test.tsx:93]: Can we add a test case here for the node count on the `fullGraph` option?

**Comment by davidsharp7** [comment]
> > So if you select etl_delivery_7_days notice how there is a grey line between public.customers and email_discounts. Given the job selected that is not on the critical path and shouldn’t appear. Simplest solution is to also filter out anything where isSelected is False.
> 
> @davidsharp7 I wanted to add to @jlukenoff response. In the `Full` graph mode, we'll want to show all dataset and job dependencies, but the connections between nodes in the graph don't need to be expanded by default. That is, they can be grouped and collapsed, allowing the user to discover and explore the graph as needed. For some insight into where I see the interaction of the graph going, see #2014

Yeah I don't think its even a binary choice as each of those views is legitimate i.e

1) I am changing dataset X I just need to see the critical path to understand impacts.
2) I am changing dataset X and I want to explore my lineage - expand/collapse
3) I want to see my entire data eco system.

**Comment by wslulciuc** [comment]
@davidsharp7 I think you make good points, and agree a more broad discussion on the graph view can be had, but I would move the discussion to an issue. Mind opening an issue to elaborate on the graph view in the UI and ways to improve the user expereince?

As for this PR, being able to toggle between a "Full" and "Critical Path" view is the main feature added, so I would merge as is with pending feedback from @phixMe. It's a start and we can follow up with improvements!

**Review by wslulciuc — APPROVED** [review]
I left some minor feedback on labeling the graph options, but otherwise 👍

**Review by phixMe — APPROVED** [review]
Thanks for adding the unit test. Looks great!

File: web/src/components/lineage/Lineage.tsx
Comment (jlukenoff) [inline web/src/components/lineage/Lineage.tsx:140]: Not sure how long we've been doing this but it looks like we've been iterating over the characters in the node id 😬 ? Update this to traverse the predecessors array and did the same for getSuccessors

**Comment by jlukenoff** [comment]
Just made a few styling tweaks that I found after adjusting to shorter labels for these configs. This should be good to merge once tests pass (I don't have access to merge yet)

**Pull request merged**
